proposals/0005-fallback-rounting.md

```diff
@@ -0,0 +1,171 @@
+- **Latency-based Routing** e.g. the next request should be served by the fastest available model in the pool. This is useful for real-time applications where latency is important.
+- **Cost-based Routing** e.g. the next request should be served by the cheapest available model in the pool
+- **Performance-based Routing** e.g. the next request should be served by the model that performs the best for my request (see GEP003).
+- etc.
+
+WRR could be useful to perform **A/B testing** of model pool setup to find the best model/params/prompt combination for a given task.
+
+### Requirements
+
+- R1: We MUST fallback to a healthy model if any in a pool doing our best to fulfill routing strategy specified in configs
+- R2: Fallback MUST be as quick as possible. No redundant wasteful retrying or waiting
+
+## Design
+
+The off-the-shelf strategy to tackle the availability issues is redundancy.
+So we have some backup models to talk to in case the main one is down.
+The mechanism of switching models in case of failures is called fallbacking.
+
+For most serious production applications that have real users, downtime is not an option. Hence, fallbacking should be
+an intrinsic feature of Glide routing despite routing strategy.
+
+There may be exceptions to this dictated by the model nature.
+For example, embedding API may require one model to be used for all embeddings produced.
+
+Based on this, we group all models we support into types (see GEP0002):
+
+- language models
+- embeddings
+- speech transcribers
+- speech synthesizers
+- etc.
+
+Each model type is represented as a list of model pools that could serve the incoming request uniformly.
+Each model pool is a list of models (along with their params & operational info like API key) and routing strategy that defines
+how incoming requests should be served.
+
+The model pool ID or name is exposed to clients and they provide IDs during requests (see GEP0002).
+This is how Glide understands which one should serve the concrete request.
+
+### Model Pools & Fallbacking
+
+TBU
+
+### Model Pool Config
+
+The config structure was already covered in [GEP0001](0001-gep.md), so in the GEP, let's just focus on the
+model pool config specifically:
+
+```YAML
+# some other configs...
+routes:
+  language:
+    - id: my-pool
+      enabled: true # # bool, true by default
+      strategy: priority # round-robin, weighted-round-robin, priority, least-latency, priority, etc.
+      models:
+        - id: openai-boring
+          enabled: true # bool, true by default
+          openai: # anthropic, azureopenai, gemini, other providers we support
+            model: gpt-3.5-turbo
+            api_key: ${env:OPENAI_API_KEY}
+            default_params:
+              temperature: 0
+```
+
+Notes:
+- The bare minimal setup must include one pool with at least one model.
+  We don't require to set up more than one model to simplify gateway setup for users who just want to try it or experiment.
```

Comment:
Given this, do we provide the second model if only one is specified?

Reply:
To simplify dev/prototyping experience it makes sense to allow to use one model. We will simply warn users that this is not a production setup. But, well, they may not need that at that in all cases. Given that, I think it doesn't make much sense to make non-prod setup hard to achieve.